### DISCUSSION TITLE
Best way to integrate with a service that already does rendering

### DISCUSSION THREAD

**Discussion by fernandocamargoai** [FitLayout/FitLayout#19] [General]
We have a project in JavaScript that already does the rendering of web pages, obtain screenshots and so on. We now want to do page segmentation using VIPS, we tried FitLayout CLI, and liked the results. However, we don't want to give up control of how we render the page, as we want to do some customizations like getting rid of cookies pop-up. I'm currently reading the documentation to better understand how we could use the REST API. From what I've seen so far, we could probably create a `Page` JSON from our rendering using puppeteer and invoke the REST API to do the segmentation from there, correct?

I'm sorry for creating an issue for my question, by the way, but the repo doesn't have discussions enabled.

**Comment by fernandocamargoai** [comment]
I'm trying to wrap my mind around the REST API. From my current understanding, I will probably need to follow this workflow:

1. Invoke puppeteer and save its output (JSON page description).
2. `POST /api/repository` to create a repository.
3. `POST /api/r/{repoId}/run/renderJson` using the output saved from puppeteer.
4. `POST /api/r/{repoId}/service` to invoke the `FitLayout.VIPS`, which will return the `AreaTree`.
5. `DELETE /api/repository/{repoId}` to delete the resources that we don't need anymore.

I'm not sure yet which Accept header I should pass in (4). Is `application/rdf+xml` the same XML format that we get when using the CLI?

Does this workflow make sense?

**Comment by fernandocamargoai** [comment]
Another question. How is the current state of the REST API? It seems that there hasn't been new releases for quite some time. Does the latest released version support VIPS? And is it production-ready?

**Comment by radkovo** [comment]
Thanks for pointing this out; I have enabled the discussions and converted the issue to a discussion.

The FitLayoutWeb REST API is bundled with a full-featured RDF storage and is primarily intended for implementing (mostly experimental) tasks that involve rendering larger sets of pages, their post-processing, which may consist of several steps (e.g. segmentation, tagging), and exporting the results (e.g. exporting selected artifacts or performing SPARQL queries).

If you just need to segment a given page using VIPS and export the results, the full FitLayoutWeb (or the fitlayout-server docker image) may be too heavy and resource consuming for you. It stores the artifacts in a persistent RDF storage, which is much slower than just doing everything in memory. As a lightweight alternative, you could just take the [VIPS demo](https://github.com/FitLayout/Demos/blob/main/src/main/java/cz/vutbr/fit/layout/demo/segmentation/SegmVIPS.java) and adapt it to your renderer:

- If you prefer to use a REST API, you can just wrap it for a lightweight Java HTTP server (e.g. make it a servlet and run it on Tomcat or another servlet container). To use your own JSON instead of the built-in renderer, you could start with the [JsonPageCreator](https://github.com/FitLayout/FitLayoutWeb/blob/main/FitLayoutWebService/src/main/java/cz/vutbr/fit/layout/web/algorithm/JsonPageCreator.java) and possibly modify it for your purpose.
- Alternatively, you can just use the existing CLI and replace the [fitlayout-puppeteer](https://github.com/FitLayout/fitlayout-puppeteer) backend with your own code. It's a JavaScript project that ultimately just outputs the JSON page description to stdout. I.e. you can modify it or implement it your way.

If some of the alternatives seems promising to you, feel free to ask more questions.

EDIT: I have also created a new release to reflect the latest changes in main. The main branch is getting updates regularly but I didn't manage to make a release for some time.

**Reply by fernandocamargoai** [reply]
Creating a lightweight Java HTTP Server with FitLayout seems to be a good solution for me. About [fitlayout-puppeteer](https://github.com/FitLayout/fitlayout-puppeteer), do you think we can install it as a NodeJS library and invoke the functions `fitlayoutDetectLines()` and `fitlayoutExportBoxes()` as in https://github.com/FitLayout/fitlayout-puppeteer/blob/3ed025034c419b3cee28938ae6328010f5b275ee/src/run.js#L178C2-L184C5?

**Reply by radkovo** [reply]
Sure you can, if you know how to do it technically. In `fitlayout-puppeteer`, I use a simple bash script to generate the client code because it actually runs in the browser, not in NodeJS. So technically, the client code isn't a standard NodeJS library right now, it's just a piece of code injected into the browser using Puppeteer. This is not nice, but I couldn't find a better solution. Anyway, feel free to reuse the code as needed, and if you have any ideas on how to improve it, I would be happy if you contribute back to the project.
